### PR TITLE
Modern build

### PR DESCRIPTION
**Pull request by Benjamin-Dobell**
Drastically simplified build thanks to recent .NET ecosystem improvements:

* Deleted `.csproj` framework variants.
* Deleted all duplicate code in the repository, _excluding_ the Unity package, which still rsyncs.
* Now taking advantage of `TargetFrameworks` to build multiple targets from the the same `.csproj`.
* Deleted all solutions tied to a particular framework.
* Now targeting `netstandard2` rather than `netstandard1.6`, which facilitated the removal of a lot of conditional compilation.
* Dropped PCL.
* Travis continuous integration is now more complete, testing both Mono and .NET Core in a Docker container.

# Motivation

Removing the duplicate code in the repository is nice (shame about Unity) and mitigates potential sync issues (right now master has code that hasn't been rsynced). The removal of the project variants/solutions certainly makes everything a lot easier to understand, so _hopefully_ should facilitate more high quality community contributions.

Modern `.csproj` syntax allows us to completely omit the source files, which simplifies things. Even better, it's easy to reference another project and re-use code, rather than for example having to rsync the test code into the runners, the runners are now just dependent on `MoonSharp.Interpreter.Tests` and it's all seamless.

Personally, I've got a bunch of things I'd like to contribute back, in particular the heap of VSCode debugger improvements; the highlight being the ability to seamlessly debug multiple scripts simultaneously in VSCode (no swapping required). These changes will make it a lot easier for me to do so.

## Motivation for targeting `netstandard2`

`netstandard2` requires a minimum .NET framework of 4.6.1, which is the same as `netstandard1.6`, so we're losing zero CLR compatibility on Windows.

In terms of Mono, the minimum supported version, if you were to for some reason run the `netstandard` build against Mono, has jumped from 4.6 to 5.4. However, in practice I very much doubt anyone was doing that seems as there were (and still are) explicit .NET 3.5 & .NET 4.0 builds that run perfectly on _really old_ Mono runtimes.

.NET core users are the only ones really affected. You now need the .NET core 2.0 runtime, where as previously you required only 1.6 (one version prior). In practice I'm doubtful this will be an issue for anyone.

Reference: https://docs.microsoft.com/en-us/dotnet/standard/net-standard

## Motivation for dropping PCL

PCL is officially deprecated, and `netstandard2` makes for a great, much more feature complete (and simpler) replacement. As above, `netstandard2` class library DLLs are _really_ portable. Given we still explicitly support .NET 3.5 and .NET 4.0 as separate build targets, I think `netstandard2` nicely covers off all other use cases where PCL previously made sense.

### CODE REVIEW

**Comment by Benjamin-Dobell** [comment]
Forgot to mention, this also helps pave the way for using newer C# language features _without_ dropping runtime support for .NET 3.5, if that's something you'd like to see, @xanathar.

This is because we now build with `msbuild` rather than `xbuild`, which was deprecated and removed from Mono quite a while ago. Using a newer Mono (with `msbuild`) also gives us access to a more recent C# compiler. However, even if we take advantage of the newer compiler by bumping the language level, we're still building against the same .NET 3.5 reference assemblies, so won't lose any compatibility[*].

* Well, some new language features take advantage of new classes/methods, we can't use those obviously unless we were to also pull in polyfills.

File: .travis.yml [hunk 1]
```diff
@@ -1,16 +1,13 @@
-language: csharp
-solution: src/moonsharp.sln
-sudo: required
+language: minimal
 
-mono: 
-  - 2.10.8
+language: minimal
```
Comment (Benjamin-Dobell) [inline .travis.yml]: This shouldn't be here. It's not causing issues, but it certainly is silly.

File: src/DevTools/PerformanceComparison/Program.cs [hunk 2]
```diff
@@ -34,59 +34,59 @@ function move(n, src, dst, via)
...
+	return true
 end
- 
+
 function Find_Solution( x )
-    for y = 1, N do
-	if Allowed( x, y ) then 
-  	    board[x][y] = true 
-	    if x == N or Find_Solution( x+1 ) then
+	for y = 1, N do
```
Comment (Benjamin-Dobell) [inline src/DevTools/PerformanceComparison/Program.cs]: Looks like the existing code was a mix of tabs and spaces and my IDE tried to make it consistent, but messed up the indentation levels.

I'll sort this out.

**Comment by admalledd** [comment]
@Benjamin-Dobell heard the news you got a few more permissions and stuff. Yay! What are the chances of digging into why this PR had to be reverted (I assume something Unity/mono/android or such caused issues?) and fixing those up? Anything I could poke at for you guys? Although, I only have VS2019 (+Linux container slaves) environments on hand to test with on the corp systems and can't install tools for the others.

On the private fork we have at work some of the stuff is only after net-standard style projects. Still working to OK publishing our whole fork instead of .PATCH files, so if you want me to help I can probably find a few hours here-and-there not irregularly but you will have to live with out github integration for a while longer.

**Comment by Benjamin-Dobell** [comment]
@admalledd Absolutely. This is definitely near the top of my list; mostly because the modern project structure makes it a lot easier for other PRs to be submitted, reviewed and merged.

It was reverted by @xanathar due to compilation issues I believe. Although unfortunately I don't know the specifics.

Basically I'm going to have to setup quite a few environments and thoroughly verify things work as expected before re-merging.

**Pull request merged**